_posts/2022-06-15-ejb-over-https.adoc

```diff
@@ -0,0 +1,340 @@
+author: cam-rod
+---
+
+The *https://github.com/wildfly-security-incubator/elytron-examples/tree/master/ejb-https[ejb-https]* example
+demonstrates how to use the Elytron subsystem to secure connections to EJBs with mutual TLS
+authentication. It also shows how to restrict usage of EJBs, implementing a FileSystem realm protected by SASL
+authentication, and accessed using a credential store. This example focuses on the TLS configuration, so if you wish to
+learn more you can refer to blog posts on
+https://wildfly-security.github.io/wildfly-elytron/blog/advanced-ejb-security/[configuring the FileSystem realm and SASL mechanism],
+and on https://wildfly-security.github.io/wildfly-elytron/blog/ejb-over-http/[connecting with HTTP].
+
+
+== An Overview of Mutual TLS authentication
+
+In a standard TLS handshake, the server presents its certificate to the client, which then inspects the contents and
+signature. If the certificate matches one that the client trusts, or is signed by an entity that it trusts, the server's
+identity is verified and a key exchange is performed to secure the session. In mutual TLS authentication, the client also
+presents its own certificate, which the server will then use to verify the identity of the client.
+
+Using the Elytron subsystem, a WildFly server can form a mutual TLS connection with a remote client. To do so, the client
+will generate a certificate, which the server will import. The client will be configured to present the certificate when
+performing the TLS handshake, which the server will verify against its own copy.
+
+== Generating the Client Certificate
+
+In order for the client to use a certificate, it must be signed by a server-trusted entity. For the purposes of this
+demonstration, we will self-sign a certificate and then import it into the server's trust store. However, such
+certificates are generally treated as insecure since the identity cannot be externally verified. To use a certificate
+signed by a trusted Certificate Authority, take a look at this https://wildfly-security.github.io/wildfly-elytron/blog/obtaining-certificates-from-lets-encrypt-using-the-wildfly-cli/[blog post].
+
+We will begin by generating a 2048-bit key pair, and saving it to a local file with the alias `client`. Using the
+Java *keytool*, we store it in a PKCS #12 keystore:
+
+[source,shell]
+----
+$ keytool -genkeypair -alias client -keyalg RSA -keysize 2048 -validity 365 -keystore /PATH/TO/sslClient.keystore -dname "CN=client" -storepass clientKeySecret
+----
+
+Note that we have set the common name on the key to `client` (using the option `-dname "CN=client"`). The server will
+try to match the common name on the certificate to the hostname presented during the TLS handshake. If the two do not
+match, the connection will fail.
+
+Once the key pair is created, we use it to generate a self-signed certificate, and save it to a certificate file:
```

Comment:
Just a small terminology thing, the -genkeypair operation actually generates the key pair and wraps the public key into an X509 self-signed certificate and stores it in the keystore. The exportcert operation just exports that certificate to a file.

Reply:
got it, will update